dynamodb/data.go

```diff
@@ -0,0 +1,237 @@
+			if len(cvtErrs) == 0 {
+				conv.WriteRow(srcTable, spTable, spCols, cvtVals)
+			} else {
+				conv.Unexpected(fmt.Sprintf("Errors while converting data: %s\n", cvtErrs))
+				conv.StatsAddBadRow(srcTable, conv.DataMode())
+				conv.CollectBadRow(srcTable, srcSchema.ColNames, srcStrVals)
+			}
+		})
+		if err != nil {
+			conv.Stats.BadRows[srcTable] += conv.Stats.Rows[srcTable]
+			conv.Unexpected(fmt.Sprintf("Can't scan the data for table %s: %s", srcTable, err))
+		}
+	}
+	return nil
+}
+
+func scan(table string, client dynamoClient, f func(map[string]*dynamodb.AttributeValue)) error {
+	var lastEvaluatedKey map[string]*dynamodb.AttributeValue
+	for {
+		// Build the query input parameters.
+		params := &dynamodb.ScanInput{
+			TableName: aws.String(table),
+		}
+		if lastEvaluatedKey != nil {
+			params.ExclusiveStartKey = lastEvaluatedKey
+		}
+
+		// Make the DynamoDB Query API call.
+		result, err := client.Scan(params)
+		if err != nil {
+			return fmt.Errorf("failed to make Query API call for table %v: %v", table, err)
+		}
+
+		// Iterate the items returned.
+		for _, attrsMap := range result.Items {
+			f(attrsMap)
+		}
+		if result.LastEvaluatedKey == nil {
+			return nil
+		}
+		// If there are more rows, then continue.
+		lastEvaluatedKey = result.LastEvaluatedKey
+	}
+}
+
+func cvtRow(attrsMap map[string]*dynamodb.AttributeValue, srcSchema schema.Table, spSchema ddl.CreateTable, spCols []string) ([]interface{}, map[string]string, []string) {
+	var err error
+	var srcStrVals []string
+	var cvtVals []interface{}
+	var cvtErrs map[string]string
+	for i, srcCol := range srcSchema.ColNames {
+		var cvtVal interface{}
+		var srcStrVal string
+		if attrsMap[srcCol] == nil {
+			cvtVal = nil
+			srcStrVal = "null"
+		} else {
+			// Convert data to the target type.
+			cvtVal, err = cvtColValue(attrsMap[srcCol], srcSchema.ColDefs[srcCol].Type.Name, spSchema.ColDefs[spCols[i]].T.Name)
+			if err != nil {
+				cvtErrs[srcCol] = fmt.Sprintf("failed to convert column: %v", srcCol)
+			}
+			srcStrVal = attrsMap[srcCol].GoString()
+		}
+		srcStrVals = append(srcStrVals, srcStrVal)
+		cvtVals = append(cvtVals, cvtVal)
+	}
+	return cvtVals, cvtErrs, srcStrVals
+}
+
+func cvtColValue(attrVal *dynamodb.AttributeValue, srcType string, spType string) (interface{}, error) {
+	switch spType {
+	case ddl.Bool:
+		switch srcType {
+		case typeBool:
+			return *attrVal.BOOL, nil
+		}
+	case ddl.Bytes:
+		switch srcType {
+		case typeBinary:
+			return attrVal.B, nil
+		case typeBinarySet:
+			return attrVal.BS, nil
+		}
+	case ddl.String:
+		switch srcType {
+		case typeMap, typeList:
+			val, err := marshalAttrValue(attrVal)
+			if err != nil {
+				return nil, fmt.Errorf("failed to convert %v to a go struct", attrVal.GoString())
+			}
+			// If the original representation is preferred, then comment the
+			// above code and change the following line to:
+			//   b, err := json.Marshal(attrVal)
```

Comment:
Suggest giving a little more context for the decision here and maybe moving the comment to the top of the case i.e. at line 135. Maybe:
			// For typeMap and typeList, attrVal is a very verbose datastructure
                         // that contains null entries for unused type cases. We strip
                         // these out using marshallAttrValue. If it important that the Spanner
                         // values can be easily unmarshalled back to dynamodb.AttributeValue
                         // types, then replace the following five lines with just: 
		        //    b, err := json.Marshal(attrVal)
                         // but note that this will consume extra Spanner storage.